Sequences restructuring

Pwhsky opened the pull request:
WIP of restructuring `sequences.py` and `SequentialProperty`.

**TODOS**
- [x] Removing the need to call `create_action` when instancing a new `SequentialProperty`.
- [x] Move `backend.core.previous` to `properties.SequentialProperty` according to #349 
- [x] Update documentation of `backend.core` and `properties.SequentialProperty`
- [x] Tutorial with SequentialProperty

##  Changes to `properties.SequentialProperty`:
### Renamed attributes:

| Old | New|
| -------- | ------- |
| `sequence_step`  |  `sequence_index`    |
| `initialization` | `initial_sampling_rule`    |
| `current`    | `sample`    |
|`current_value`  | `sampling_rule`|

### Usage
A `SequentialProperty` can be initialized and used as:
```python
from deeptrack import SequentialProperty
from numpy.random import randn

seq_prop = SequentialProperty(
     initial_sampling_rule=0,
     sampling_rule=lambda: randn(),
     sequence_index=0,
     sequence_length=10,
)
print(seq_prop.sample()) # Sample with the sampling rule.

'-0.4214685768'
```

* Added two new methods: `set_sequence_length()`, `set_sequence_index()`, which lets you directly set the Property.sequence_length and Property.sequence_index attributes without having to use `SequentialProperty.store()`.

* Three new optional argument when creating a `SequentialProperty`:
 `sampling_rule`, `sequence_length`,  `sequence_index`.


## Changes to `features.Feature`:
`to_sequential()` is a new `Feature` class method intended to replace the `Sequential()` method found in `sequences.py`.

### Usage
`Sequential()` is used as:
```
rotating_ellipse = deeptrack.sequences.Sequential(rotation=get_rotation)
imaged_rotating_ellipse = optics(rotating_ellipse)
```
The new `Feature` method `to_sequential()` is used as:
```
rotating_ellipse = ellipse.to_sequential(rotation=get_rotation)
imaged_rotating_ellipse = optics(rotating_ellipse)
```

Pwhsky committed:
features: added to_sequential()

Pwhsky committed:
refactoring: SequentialProperty

Pwhsky committed:
typo in type hint

Pwhsky committed:
added dependency support for the set methods.

Pwhsky committed:
docs for to_sequential()

Pwhsky committed:
Update features.py

Pwhsky committed:
Update features.py

Pwhsky committed:
Merge branch 'DeepTrackAI:develop' into sequences-restructuring

Pwhsky committed:
Merge pull request #344 from Pwhsky/sequences-restructuring

Migrates branch from private fork to DeepTrack2 AL/sequences-restructuring branch.

Pwhsky committed:
Add files via upload

giovannivolpe committed:
Merge branch 'develop' into AL/sequences-restructuring

giovannivolpe commented:
TODO list:
- [x] add example for `SequentialPropery` in DTAT306_properties.ipynb

giovannivolpe committed:
edits

Update sequences.py

Update properties.py

Update sequences.py

Update sequences.py

Update DTAT341_sequences.ipynb

giovannivolpe committed:
Update test_properties.py

giovannivolpe committed:
Update properties.py

giovannivolpe committed:
u

giovannivolpe committed:
Update properties.py

giovannivolpe committed:
Update test_properties.py

giovannivolpe committed:
Update properties.py

giovannivolpe committed:
Update __init__.py

giovannivolpe committed:
Update properties.py

giovannivolpe committed:
Update properties.py

giovannivolpe committed:
Update features.py

giovannivolpe committed:
Update features.py

giovannivolpe committed:
Update sequences.py

Pwhsky committed:
documentation and type hints

Pwhsky committed:
import Optional, Dict

Pwhsky committed:
missing comma...

Pwhsky committed:
current attributes

Pwhsky committed:
fix: typo in docstring.

Pwhsky committed:
documentation for sequences

Pwhsky committed:
typo

Pwhsky committed:
Update sequences.py

Pwhsky committed:
:

Pwhsky committed:
future import

Pwhsky commented:
> TODO list:
> 
> * [x]  add example for `SequentialPropery` in DTAT306_properties.ipynb

@giovannivolpe Added.

Pwhsky committed:
documentation

Pwhsky committed:
DTAT306: Added SequentialProperty Example. (#350)

* Added SequentialProperty example

* typo

Pwhsky committed:
Merge branch 'develop' into AL/sequences-restructuring

Pwhsky committed:
set length & index in seqprop __init__

Pwhsky committed:
variable rename: sequence_step <- sequence_index

Pwhsky committed:
variable rename: sequence_step <- sequence_index

Pwhsky committed:
Update test_properties.py with new variable name

Pwhsky committed:
update documentation of SequentialProperty

Pwhsky commented:
Hi @BenjaminMidtvedt!
I am proposing some attribute name changes to `SequentialProperty`, and your input is very welcome as to what could be optimal in terms of interpretability. 

The current ideas are:

### Attributes
`initialization` changed to `initial_sampling_rule` (implemented)
`current` changed to `sample` (implemented)
 
This should make a more clear separation between a *sampling rule* in a SequentialProperty and the stored *values* which are added with `.store()` during the sequence evolution/evaluation.

Pwhsky committed:
Variable name changes

Pwhsky committed:
variable name change: current and initialization

Pwhsky committed:
Update test_properties.py

Pwhsky committed:
Update features.py

Pwhsky committed:
Update sequences.py

Pwhsky committed:
sampling rule in init.

Pwhsky committed:
attributes

Pwhsky committed:
Update properties.py

Pwhsky committed:
Update features.py

Pwhsky committed:
Update properties.py

Pwhsky committed:
Update test_sequences.py

Pwhsky committed:
Revert "Update test_sequences.py"

This reverts commit 503ea57bf4fa3f3937a601145094383d60e7e819.

Pwhsky committed:
Update properties.py

Pwhsky committed:
Update test_sequences.py

Pwhsky committed:
Update test_sequences.py

Pwhsky committed:
sequentialproperty docs

Pwhsky committed:
sequentialproperty docs

Pwhsky committed:
markov chain example in properties tutorial

Pwhsky committed:
markov chain example in properties tutorial

Pwhsky committed:
Update seq_prop names

Pwhsky committed:
new sequentialproperty attribute names

Pwhsky committed:
Update features.py

Pwhsky committed:
Update properties.py

Pwhsky committed:
Update properties.py

Pwhsky committed:
Update properties.py

Pwhsky committed:
Update properties.py

Pwhsky committed:
Update properties.py

Pwhsky committed:
Update properties.py

Pwhsky committed:
removed duplicate method

Pwhsky committed:
removed duplicate method

Pwhsky committed:
added example in docs

Pwhsky committed:
added example in docs

Pwhsky committed:
Updated tutorial with new names and examples

Pwhsky committed:
Updated tutorial with new names and examples

Pwhsky committed:
Update unit test

Pwhsky committed:
Update unit test

Pwhsky committed:
Update test_properties.py

Pwhsky committed:
Update test_properties.py

Pwhsky committed:
Merge branch 'develop' into AL/sequences-restructuring

Pwhsky committed:
docs: example of to_sequential()

Pwhsky committed:
create_action in __init__

Pwhsky committed:
Added Previous() to SequentialProperty

Pwhsky committed:
Removed previous from backend.core

Pwhsky committed:
removed upsample from unit tests

Pwhsky committed:
Specific imports

Pwhsky committed:
u

Pwhsky committed:
sequences type hints

Pwhsky committed:
Updated SequentialProperty example

Pwhsky committed:
comma

giovannivolpe committed:
Merge branch 'gv/final/features' into AL/sequences-restructuring

giovannivolpe committed:
Merge branch 'AL/sequences-restructuring' of https://github.com/DeepTrackAI/DeepTrack2 into AL/sequences-restructuring

giovannivolpe committed:
Update __init__.py

giovannivolpe committed:
Update features.py

giovannivolpe committed:
Update DTAT306_properties.ipynb

giovannivolpe committed:
Update DTAT306_properties.ipynb

giovannivolpe committed:
Update DTAT341_sequences.ipynb

giovannivolpe committed:
Update DTAT306_properties.ipynb

giovannivolpe committed:
Update properties.py

giovannivolpe committed:
Update test_properties.py

giovannivolpe committed:
Update test_properties.py

giovannivolpe committed:
Update features.py

giovannivolpe committed:
Update features.py

giovannivolpe committed:
Update optics.py

giovannivolpe committed:
Update sequences.py

giovannivolpe committed:
Update properties.py

giovannivolpe committed:
Update test_sequences.py

giovannivolpe committed:
Update features.py
